lib/Client.php

```diff
@@ -521,23 +523,24 @@ private function write(string $data, bool $close = false): Promise {
             $this->close();
         }
 
+        $this->timeoutCache->renew($this->id);
+
         return new Success;
     }
 
     /**
      * Attempts to write the given data directly to the client socket. Returns any unwritten data.
      *
      * @param string $data Data to write.
+     * @param bool $strict If `true`, close the client if no bytes are able to be written.
      *
      * @return string Remaining unwritten data.
      *
      * @throws \Aerys\ClientException If the client has disconnected.
      */
-    private function send(string $data): string {
+    private function send(string $data, bool $strict = false): string {
         $bytesWritten = @\fwrite($this->socket, $data);
-        if ($bytesWritten === false
-            || ($bytesWritten === 0 && (!\is_resource($this->socket) || @\feof($this->socket)))
-        ) {
+        if ($bytesWritten === false || ($strict && $bytesWritten === 0)) {
```

Comment:
What if the write buffer is full? Then it will just close the connection?

Reply:
Yep, writing this style of code is only fine in an onWritable callback as there we're guaranteed free space in the write buffer.

Reply:
`$strict` is only set to true in the call from the `onWritable` callback.

Reply:
Then I recommend calling this `$fromOnWritable` to explicitly hint that it needs to be called from within onWritable